design/open/high-level-python-wrapper.md

```diff
@@ -0,0 +1,314 @@
+    def path(self) -> self: ...
+    def modified_time(self) -> datetime.datetime: ...
+    def size_bytes(self) -> int: ...
+    def content_type(self) -> Optional[str]: ...
+    def metadata(self) -> dict[str, str]: ...
+    def physical_address(self) -> str: ...
+    def delete(self): ...
+
+
+class Change(NamedTuple):
+    type: Literal(['added', 'removed', 'changed', 'conflict,', 'prefix_changed'])
+    path: str
+    path_type: Literal(['common_prefix', 'object'])
+    size_bytes: Optional[int]
+
+
+class ServerConfiguration:
+    def __init__(self, client: Client = DefaultClient): ...
+    def version(self) -> str: ...
+    def storage_config(self) -> ServerStorageConfiguration: ...
+
+
+class ServerStorageConfiguration(NamedTuple):
+    blockstore_type: str
+    pre_sign_support: bool
+    import_support: bool
+
+
+class ImportManager:
+    def __init__(self, repository_id: str, reference_id: str, client: Client = DefaultClient): ...
+    def start(self) -> str:
+        'start import, reporting back (and storing) a process id'
+        ...
+    def wait(self, poll_interval: timedelta = timedelta(seconds=2)) -> ImportResult:
+        'poll a started import task ID, blocking until completion'
+        ...
+    def run(self, poll_interval: Optional[timedelta] = None) -> ImportResult:
+        'same as calling start() and then wait()'
+        ...
+
+
+class ImportResult(NamedTuple):
+    commit: Commit
+    ingested_objects: int
+
+class BranchTransaction(WriteIOBase):
+    def __init__(self, repository_id: str, source_branch_id: str, client: Client = DefaultClient): ...
+    
+    def __enter__(self):
+        'Create ephemeral branch from the source branch (e.g. <source_branch_id>-txn-<uuid>'
+        ...
+
+    def __exit__(self, type, value, traceback):
+        """if successful: commit, merge, delete ephemeral branch
+           otherwise, leave branch and report with a meaningful error
+        """
+        ...
+
+```
+
+While this list is fairly exhaustive, it might require a few additional tweaks and additions.
+
+Additionally, we define the following exception hierarchy:
+
+```python
+
+# lakefs.exceptions
+class LakeFSException(Exception):
+    status_code: int
+    message: str
+
+class NotFoundException(LakeFSException): ...
+class NotAuthorizedException(LakeFSException): ...
+class ServerException(LakeFSException): ...
+class UnsupportedOperationException(LakeFSException): ...
+```
+
+Other, more specific exceptions may subclass these, but all errors returned by the lakeFS server should sub-class one of these to make error handling easier for developers.
+
+The only exception should be errors returned by functions that, for compatibility should return specific error types. (this might be true for e.g. `open()` and errors returned by the `TextIO` or `BinaryIO` implementations).
+
+## Higher Level Utilities
+
+### I/O - reading/writing objects
```

Comment:
Without going into too much of the semantic details, but this could have significant overlap with the abstractions provided by [`fsspec`](https://filesystem-spec.readthedocs.io/en/latest/features.html#file-buffering-and-random-access).

We've built on top of these in our [`lakefs-spec` library](https://github.com/appliedAI-Initiative/lakefs-spec), see the [`LakeFSFile` class](https://github.com/appliedAI-Initiative/lakefs-spec/blob/main/src/lakefs_spec/spec.py#L604) in particular. There might be some potential for reuse/alignment here.

Reply:
I agree! This is one of the reasons that Nico and I discussed as to why you should be a reviewer and/or contributor to this:

1. It would make sure that the lakeFS configuration is consistent whether you're using this client or the fsspec impl. (in terms of environment variables, yaml config, etc)
2. IO behavior is consistent across lakeFS clients - lakectl, this sdk, fsspec, etc - all perform the same way when it comes to pre-signed URLs, direct access and so on.

The easiest way to ensure those is to use shared code where possible and avoid multiple implementations of the same logic.